HLSL: Support depth comparison texture sampling in SM 2/3.

rdb opened the pull request:
In DX9, all samples of a depth texture result in a depth comparison (in fact, disabling this requires use of the special `INTZ` format, only available on DX10+ hardware).  This works with all 4-component texture sample functions (tex2Dlod, tex2Dbias, tex2Dproj).

This PR provides access to this functionality by enabling use of the Dref texture sample instruction and passing in the depth compare value into the appropriate register.  It assumes that there is indeed a depth texture bound to the appropriate register (which is the application's responsibility, and not something we can check).

Even though the regular `texld` instruction does support this as well, HLSL does not (to my knowledge) allow access to this variant since `tex2D` only accepts 2 coordinates.  To work around this, this PR turns an OpImageSampleDrefImplicitLod into a `tex2Dproj` with a w coordinate of 1.0, which is effectively equivalent.

Thanks for your consideration!

HansKristian-Work reviewed (changes requested):
I'd like to see a test here to see what is going on.

rdb commented:
I added a test for all the samplers supported by SM 3.0.

As you can see it's just a matter of passing the dref into the third coordinate.  I verified that the resulting shader works correctly for shadow mapping on both Intel and NVIDIA hardware, and furthermore, this is also what the Cg compiler is producing when using a shadow sampler and targeting SM 3.0 (aside from the fact that it can directly generate a `texld` instead of `texldp`, but that has no effect on the result).

Thanks!

HansKristian-Work commented on reference/shaders-hlsl/frag/tex-sampling.sm30.frag:
```diff
@@ -0,0 +1,87 @@
+    float4 _216 = float4(texCoord2d, 0.0f, 2.0f);
+    float4 _219 = _216;
+    _219.z = _216.w;
+    texcolor.w += tex2Dproj(tex2dShadow, float4(_219.xy, _216.z, _219.z));
```
This looks weird. What is the return type of tex2Dproj here? I would guess float4, in which case, we need to emit a `.x` swizzle after the expression since Dref is expected to be scalar in SPIR-V.

rdb replied:
Good catch, I'll change that.

HansKristian-Work commented on spirv_hlsl.cpp:
```diff
@@ -2994,35 +2993,45 @@ void CompilerHLSL::emit_texture_op(const Instruction &i, bool sparse)
-				coord_filler += ", 0.0";
-			coord_expr = "float4(" + coord_expr + coord_filler + ", " + to_expression(lod) + ")";
-			modifier_count++;
+				coord_expr += ", " + to_expression(dref);
```
Should dref be duplicated twice out to 3 components when using tex1D?

HansKristian-Work replied:
I guess it doesn't hurt though, but ...

rdb replied:
Not necessary, just matching the behaviour of the NV Cg compiler, which put the depth in both y and z.  My guess was that it might have been one less instruction to just swizzle than having to add in a 0?  No idea.  Should I change it?

HansKristian-Work replied:
I would normally expect the second coordinate to be 0 for 1D, but if Cg does it that way I don't mind it being there as-is. Just add a comment I guess to not confuse future readers.

HansKristian-Work reviewed (changes requested):
Also missing `reference/opt/shaders-hlsl/frag/tex-sampling.sm30.frag` to be committed.

rdb commented:
I just went ahead and just changed the second coordinate to 0.0, because it's clearer.  I verified that the resulting shader still works correctly with a 1D depth map.

I also added a return swizzle and added the optimized reference file as requested.

HansKristian-Work reviewed (approved):
LGTM, thanks.
